tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py

```diff
@@ -0,0 +1,248 @@
+    assert abs(teams[0].cumulated_rating - teams[1].cumulated_rating) < 50
+
+
+def test_make_teams_with_full_sized_search(player_factory):
+    s = make_searches([925, 1084, 1015, 1094, 1403, 1526, 1637, 1718], player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop(), s.pop(), s.pop()])
+    s.append(c1)
+
+    team_a = CombinedSearch(*[c1])
+    team_b = CombinedSearch(*[s[0], s[1], s[2], s[3]])
+    teams = TeamMatchMaker(4).make_teams(s)
+
+    assert set(teams[0].players) == set(team_a.players)
+    assert set(teams[1].players) == set(team_b.players)
+
+
+def test_make_teams_with_almost_full_sized_search(player_factory):
+    s = make_searches([1415, 125, 1294, 584, 1303, 1526, 1237, 1218], player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    s.append(c1)
+
+    team_a = CombinedSearch(*[c1, s[3]])
+    team_b = CombinedSearch(*[s[0], s[1], s[2], s[4]])
+    teams = TeamMatchMaker(4).make_teams(s)
+
+    assert set(teams[0].players) == set(team_a.players)
+    assert set(teams[1].players) == set(team_b.players)
+
+
+def test_handle_impossible_team_splits(player_factory):
+    s = make_searches([1415, 125, 1294, 584, 1303, 1526, 1237, 1218], player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c2 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c3 = CombinedSearch(*[s.pop(), s.pop()])
+    s.append(c1)
+    s.append(c2)
+    s.append(c3)
+
+    with pytest.raises(UnevenTeamsException):
+        TeamMatchMaker(4).make_teams(s)
+
+
+def test_ignore_impossible_team_splits(player_factory):
+    s = make_searches([1415, 125, 1294, 584, 1303, 1526, 1237, 1218], player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c2 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c3 = CombinedSearch(*[s.pop(), s.pop()])
+    s.append(c1)
+    s.append(c2)
+    s.append(c3)
+
+    matches = TeamMatchMaker(4).find(s)
+
+    assert len(matches) == 0
+
+
+def test_game_quality(player_factory):
+    s = make_searches([100, 100, 100, 100, 100, 100, 100, 100], player_factory)
+    team_a = CombinedSearch(*[s[0], s[1], s[2], s[3]])
+    team_b = CombinedSearch(*[s[4], s[5], s[6], s[7]])
+    game = TeamMatchMaker(4).calculate_game_quality((team_a, team_b))
+
+    assert game.quality == 1.0
+
+
+def test_game_quality_2(player_factory):
+    s = make_searches([100, 100, 400, 400, 400, 400, 100, 100], player_factory)
+    team_a = CombinedSearch(*[s[0], s[1], s[2], s[3]])
+    team_b = CombinedSearch(*[s[4], s[5], s[6], s[7]])
+    game = TeamMatchMaker(4).calculate_game_quality((team_a, team_b))
+
+    assert game.quality == 0.5
+
+
+def test_game_quality_3(player_factory):
+    s = make_searches([100, 100, 4000, 4000, 4000, 4000, 100, 100], player_factory)
+    team_a = CombinedSearch(*[s[0], s[1], s[2], s[3]])
+    team_b = CombinedSearch(*[s[4], s[5], s[6], s[7]])
+    game = TeamMatchMaker(4).calculate_game_quality((team_a, team_b))
+
+    assert game.quality == 0.0
```

Comment:
Where are the tests for the private methods!? You know, like the one that had a bug because it was using negative indices. That could have been found with a unit test or two.

Reply:
imfine

Reply:
I'll add some